components/2.0/PasswordConnections.qml

```diff
@@ -0,0 +1,151 @@
+
+    // if provided supposed to have text property
+    property var errMsg: QtObject
+    property var txtMsg: QtObject
+
+    // gets focus when password renewal dialog closes
+    property var getsBackFocus: QtObject
+
+    // items which are disabled when password dialog is active
+    property var disabledItems: []
+
+    property var itemStates: [] // bool
+
+    TextConstants { id: textConstants }
+
+    function clearPwd() {
+        if(typeof pwdItem.password !== "undefined")
+            pwdItem.password = ""
+        else if(typeof pwdItem.text !== "undefined")
+            pwdItem.text = ""
+    }
+
+    // disable other selected Main.qml items
+    // when password dialog is active (visible)
+    function toggleDisabledItems(disable) {
+	var i;
+
+        if(disable) {
+            // save enabled state of items
+            for(i=0; i < disabledItems.length; i++) {
+                itemStates[i] = disabledItems[i].enabled
+                disabledItems[i].enabled = false
+            }
+        } else {
+            // restore original enabled values
+            for(i=0; i < disabledItems.length; i++)
+                disabledItems[i].enabled = itemStates[i]
+        }
+    }
+
+    // tell greeter we handle expired passwords,
+    // default is no password renewal in old themes
+    Component.onCompleted: sddm.enablePwdRenewal()
+
+    // handles password renewal events
+    Connections {
+
+        target: renewalDialog
+
+        onOk: {
+            sddmProxy.pamResponse(renewalDialog.password)
+        }
+
+        onCancel: {
+            sddmProxy.cancelPamConv()
+        }
+
+        onError: {
+            // unlikely: invalid prompt,
+            // i.e. new/repeat prompt for dialog not found in request
+            if(errMsg.objName !== "")
+                errMsg.text = msg
+        }
+
+        onVisibleChanged: {
+            if(!renewalDialog.visible) {
+                renewalDialog.clear() // clear pam infos
+                clearPwd()
+
+                // last selected user or input field gets focus back
+                getsBackFocus.forceActiveFocus()
+            }
+            // disable specified Main.qml items during password renewal,
+            // e.g. user list, input fields etc.
+            toggleDisabledItems(renewalDialog.visible)
+        }
+    }
+
+    Connections {
+
+        target: sddmProxy
+
+        onLoginSucceeded: {
+            // ...will not be reached as greeter stops after successfull login...
+        }
+
+        onLoginFailed: {
+            if(txtMsg.objName !== "")
+            {
+                txtMsg.color = "red"
+                txtMsg.text = textConstants.loginFailed
+                // filter out login failure details
+                if(err_msg != "Authentication failure" &&
+                   err_msg != "Password change aborted." &&
+                   err_msg != "Authentication token manipulation error")
+                   txtMsg.text += "\n" + err_msg
```

Comment:
first:
> There's a very important comment about reading PAM data and i18n. That needs fixing at a core level.

well, i agree the code should be ready for i18n setups.
if we have to handle multi-language messages from PAM itself, then the PamBackend.cpp has first to be amended. it only handles ascii messages currently. i based my changes on that version to extend it and use those messages in the frontend. unfortunately pam conversation gives no kind of message ID to easily identify messages but just msg_style and raw string which is not enough. at the moment the type is just found my string comparison. who will do that i18n extensions the to pam backend? as we need this here to maybe i can implement, but not sure if i have time soon.

> I've had this before when doing multiple auth prompts. Probably best done in the C++ code rather than the QML.

we can do some preprocessing outside of qml, sure.